deploy/tests/test_yaml_api.py

```diff
@@ -0,0 +1,402 @@
+    file = tmp_path / 'GHOST_FILE'
+    file.exists().should.equal(False)
+    return file
+
+@pytest.fixture()
+def firesim_argparser():
+    return firesim.construct_firesim_argparser()
+
+@pytest.fixture()
+def fs_cl2args(firesim_argparser):
+    """fixture that bundles calling parse_args and shlex.split to abbreviate usage in tests"""
+    return lambda x: firesim_argparser.parse_args(shlex.split(x))
+
+@pytest.mark.usefixtures("aws_test_credentials")
+class TestConfigBuildAPI:
+    """ Test config_{build, build_recipes}.yaml APIs """
+
+    # run_test("invalid-build-section")
+    @patch('firesim.buildafi')
+    def test_invalid_build_section(self, buildafi_mock, build_yamls, fs_cl2args):
+        # @patch modifies the symbol table with our Mock but we're dispatching through
+        # a dict and that needs to be patched as well, otherwise our mock won't be called but
+        # the real function will be
+        firesim.BUILD_TASKS['buildafi'] = buildafi_mock
+        os.environ['FIRESIM_SOURCED'] = '1'
+        # at the beginning of the test build_yamls contains the backup-sample-configs
+        # but we can show exactly what we're doing different from the default by
+        build_yamls.build.load(dedent("""
+            builds:
+
+            agfis-to-share:
+                - testing-recipe-name
+
+            share-with-accounts:
+                INVALID_NAME: 123456789012
+            """))
+
+        build_yamls.write()
+        args = fs_cl2args(' '.join(['buildafi']+build_yamls.args))
+        pprint(build_yamls.build.data)
+        firesim.main.when.called_with(args).should.throw(TypeError)
+        buildafi_mock.assert_not_called()
+
+
+    @patch('firesim.buildafi')
+    def test_programmatic_invalid_build_section(self, buildafi_mock, build_yamls, firesim_argparser):
+        # @patch modifies the symbol table with our Mock but we're dispatching through
+        # a dict and that needs to be patched as well, otherwise our mock won't be called but
+        # the real function will be
+        firesim.BUILD_TASKS['buildafi'] = buildafi_mock
+        os.environ['FIRESIM_SOURCED'] = '1'
+        # at the beginning of the test build_yamls contains the backup-sample-configs
+        # but we can show exactly what we're doing different from the default by
+        # by programatically modifying from the sample to remove all the builds
+        #build_yamls.build.data['builds'] = []
+        # lol sometimes, it's hard to tell exactly how the yaml will get loaded because None is what you get
+        build_yamls.build.data['builds'] = None
+        build_yamls.write()
+        args = firesim_argparser.parse_args(['buildafi'] + build_yamls.args)
+        pprint(build_yamls.build.data)
+        firesim.main.when.called_with(args).should.throw(TypeError)
+        buildafi_mock.assert_not_called()
+
+
+    # run_test("invalid-aws-ec2-inst-type")
+    @pytest.mark.skip(reason="unclear as to what this is testing even in Abe's impl")
```

Comment:
I wanted to see if the AWS EC2 API would error out saying invalid instance type. We can probably remove this test since it is just testing if the AWS EC2 API can properly catch its own errors.

Reply:
Ok.  Thanks for clarifying.  I wasn't sure if you were somehow validating that in the YAML early or testing AWS.  I'll remove that.

Reply:
I added a test to test_awstools.py that shows `moto` will let you create instances with bogus types when using the mocking library.  I might also add a test for whether aws will properly catch something bogus as an example of how to write a test that does real AWS stuff.  Checking for that case isn't horrible since we are allowing something that a user puts in the config file to go directly to AWS without any kind of validation first.